### ISSUE TITLE
bug in handshake detection and conversion to hccapx

### ISSUE THREAD

**Issue by ZerBea** [aircrack-ng/aircrack-ng#1993]
Detection of handshakes is buggy, because aircrack-suite doesn't take care about replay count.
Conversion to hccapx is buggy, because aircrack-ng doesn't take care about replay count, but set message_pair to replay count checked.

Message pair field:
https://hashcat.net/wiki/doku.php?id=hccapx

Attached example:
replay count M1 = 2
replay count M2 = 3 (not matching to M1)

./aircrack-ng -j test actest.pcap convert the handshake to hccapx and set message_pair to 0 (replay count checked) - should be 0x80 (replay count not checked)

$ wlanhcxinfo -i test.hccapx -a -A -s -S -R -p
00789efb3439:7db254a022c7742be0e4511346265d20c3f20e3f11b042e4c4dc7b118b92ea19:74d21d497796:e6f037d7323966bea2bff97f778b266c07bdfc96960869f857ed17f638cca8a9:0000000000000003:00

[actest.pcap.zip](https://github.com/aircrack-ng/aircrack-ng/files/2613004/actest.pcap.zip)

**Comment by ZerBea** [comment]
let me give you another example:

[actest2.pcap.zip](https://github.com/aircrack-ng/aircrack-ng/files/2613015/actest2.pcap.zip)

$ wlanhcxinfo -i test.hccapx -a -A -s -S -R -p
00789efb3439:7db254a022c7742be0e4511346265d20c3f20e3f11b042e4c4dc7b118b92ea1a:74d21d497796:e6f037d7323966bea2bff97f778b266c07bdfc96960869f857ed17f638cca8a9:0000000000000003:00
ANONCE is taken from M3
EAPOL is taken from M2

Message pair field should be 0x02 (if you checked replay count) or 0x82 (if you didn't check replay count) instead of 0!

The message pair field is used to control hashcat's nonce error corrections behaviour and JtR's nonce error corrections behaviour, if converted to this handshake format. So it's important to set correct values here. Otherwise the handshake is uncrackable.

This analysis is based on wpa-sec data.
If you need more informations, please let me know.

**Comment by ZerBea** [comment]
Also, aircrack-ng irgnore EAPOL timer values:
In this case M1 (replay count 1) from 05:59:18 and M2 (replay count 2) from 07:36:11
is detected as valid handshake and converted to hccapx.
The handshake (not the PMKID) is only crackable running a high nonce-error correction, but aircrack-ng convert it to hccapx and set this message pair value to 0 and .

[actest3.pcap.zip](https://github.com/aircrack-ng/aircrack-ng/files/2619216/actest3.pcap.zip)

More information here:
https://www.researchgate.net/publication/3434048_Effects_of_the_EAPOL_timers_in_IEEE_8021X_authentication
ore here:
https://community.cisco.com/t5/wireless-mobility-documents/eap-timers-on-wireless-lan-controllers/ta-p/3122477

**Comment by ZerBea** [comment]
Now we make it complete and run the pcap through wpaclean:
$ ./wpaclean ac4cleaned.pcap ac4test.pcap
Pwning ac4test.pcap (1/1 100%)
Net 18:a6:f7:36:c2:b9 TurkTelekom_TC2B9

[ac4cleaned.pcap.zip](https://github.com/aircrack-ng/aircrack-ng/files/2619725/ac4cleaned.pcap.zip)

Now we have a completely uncrackable handshake and a completely uncrackable hccapx file. But aircrack-ng-suite made valid(!) hashes from them. Most of the aircrack-ng users will upload this (useless) files to forums or ask for help to recover the passwords.

If you have some questions, please feel free to ask. I really like the aircrack suite (one reason for me to decide not to implement wep).

BTW:
Beacons are nice and easy to capture.
But more and improtant information ara in proberesponse frames, authentication frames and association frames. Plese do not filter them out.

**Comment by jbenden** [comment]
The commit d77e678 adds very basic support for the message_pair field.

This simply works around problems (hopefully) for now. It does not check timers or replay counts.

**Comment by ZerBea** [comment]
Great. That is a good first workaround.
A major problem is the combination of aireplay-ng and airodump-ng. I noticed that aireplay-ng transmits too much deauthentications and that cause the AP to reset it's EAPOL timer. Now the AP reacts as follows:
a) increment replaycounter
b) increment last byte of ANONCE
c) a combination of a) and b)

That happens very often (based on wpa-sec data)

May workaround against this:
- the attacktool (hcxdumptool) asks the AP to retransmit his M1 (should be easy to implement that in besside-ng and aireplay-ng) to make sure we have no packet loss
- the conversion tool (hcxpcaptool)
  + measures the time between the Mx messages
  + checks the replaycount
  + compare at least 2 ANONCEs from M1 or 2 ANONCEs from M3 to detect a packet loss (and activate JtRs nonce-fuzzing or hashcats nonce-error-corrections)

**Comment by ZerBea** [comment]
BTW:
When converting to hccapx, it's important that you take EAPOL only from M2 or (not zeroed) M4 and NONCE from M1 or M3.
If you take EAPOL from M3, we are not able to convert the hccapx back to a cap.

**Comment by jbenden** [comment]
Thanks!

I've added a warning message when creating a HCCAPX file, with M3 EAPOL content. So it will still create the file; however, it'll complain about doing so. This change was made at 1181655c71e42e96a0070d4b52acf1939e35136a.

**Comment by ZerBea** [comment]
Please keep also in mind that an EAPOL M3 can be greater than 255 bytes. The size depends on the encrypted RSN information of the AP. hccapx isn't able to handle this.
I talked to Atom and Magnum about this problem and we decided to create a new format:
https://github.com/hashcat/hashcat/issues/1816

**Comment by jbenden** [comment]
I've added a block of code to emit a warning on the data exceeding the size, during export to hccapx. However, for reference, currently `aircrack-ng` silently discards M3 messages that EAPOL data exceeds 256 bytes.

**Comment by ZerBea** [comment]
Yes, I do the same on M3 (and not SNONCE zeroed M4), but I hope, we find a good hccapx successor, soon.
